File: Problem-041/solution.js
function pandigitalPrime(n) {
    const digits = [];

    for (let i = n; i >= 1; i--) {
        digits.push(i);
    }

    return findLargestPrime(digits, 0);
}


function findLargestPrime(arr, index) {
    if (index === arr.length) {
        const num = Number(arr.join(""));

        return isPrime(num) ? num : -1;
    }

    for (let i = index; i < arr.length; i++) {
        [arr[index], arr[i]] = [arr[i], arr[index]];

        const result = findLargestPrime(arr, index + 1);
        [arr[index], arr[i]] = [arr[i], arr[index]];

        if (result !== -1) {
            return result;
        }
    }

    return -1;
}


function isPrime(num) {
    if (num < 2) return false;

    for (let i = 2; i * i <= num; i++) {
        if (num % i === 0) return false;
    }

    return true;
}


console.log(pandigitalPrime(4)); 

File: Problem-041/solution.py
from itertools import permutations
from math import isqrt


def is_prime(num):
    if num < 2:
        return False

    for i in range(2, isqrt(num) + 1):
        if num % i == 0:
            return False

    return True


def pandigital_prime(n):
    # Create digits in descending order: n, n-1, ..., 1
    digits = [str(i) for i in range(n, 0, -1)]

    # Generate permutations
    for perm in permutations(digits):
        num = int("".join(perm))

        # First prime found is the largest one
        if is_prime(num):
            return num

    return -1


print(pandigital_prime(4))
